core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java

```diff
@@ -0,0 +1,180 @@
+
+/**
+ * The AnalysisModule is the main extension point for node and index level analysis components. The lucene classes
+ * {@link org.apache.lucene.analysis.Analyzer}, {@link org.apache.lucene.analysis.TokenFilter}, {@link org.apache.lucene.analysis.Tokenizer}
+ * and {@link org.apache.lucene.analysis.CharFilter} can we extended in plugins and registered on node startup when the analysis module
+ * gets loaded. Since elasticsearch needs to create multiple instances for different configurations dedicated factories need to be provided for
+ * each of the components:
+ * <ul>
+ *     <li> {@link org.apache.lucene.analysis.Analyzer} can be exposed via {@link AnalyzerProvider} and registered on {@link #addAnalyzer(String, AnalysisProvider)}</li>
+ *     <li> {@link org.apache.lucene.analysis.TokenFilter} can be exposed via {@link TokenFilterFactory} and registered on {@link #addTokenFilter(String, AnalysisProvider)}</li>
+ *     <li> {@link org.apache.lucene.analysis.Tokenizer} can be exposed via {@link TokenizerFactory} and registered on {@link #addTokenizer(String, AnalysisProvider)}</li>
+ *     <li> {@link org.apache.lucene.analysis.CharFilter} can be exposed via {@link CharFilterFactory} and registered on {@link #addCharFilter(String, AnalysisProvider)}</li>
+ * </ul>
+ *
+ * The {@link org.elasticsearch.indices.analysis.AnalysisModule.AnalysisProvider} is only a functional interface that allows to register factory constructors directly like the plugin example below:
+ * <pre>
+ *     public class MyAnalysisPlugin extends extends Plugin {
+ *       @Override
+ *       public String name() {
+ *         return "analysis-my-plugin";
+ *       }
+ *
+ *       @Override
+ *       public String description() {
+ *         return "my very fast and efficient analyzer";
+ *       }
+ *
+ *       public void onModule(AnalysisModule module) {
+ *         module.addAnalyzer("my-analyzer-name", MyAnalyzer::new);
+ *       }
+ *     }
+ * </pre>
+ */
+public final class AnalysisModule extends AbstractModule {
+
+    private final Environment environment;
+    private final Map<String, AnalysisProvider<CharFilterFactory>> charFilters = new HashMap<>();
+    private final Map<String, AnalysisProvider<TokenFilterFactory>> tokenFilters = new HashMap<>();
+    private final Map<String, AnalysisProvider<TokenizerFactory>> tokenizers = new HashMap<>();
+    private final Map<String, AnalysisProvider<AnalyzerProvider>> analyzers = new HashMap<>();
+    private final Map<String, org.apache.lucene.analysis.hunspell.Dictionary> knownDictionaries = new HashMap<>();
+
+    /**
+     * Creates a new AnalysisModule
+     */
+    public AnalysisModule(Environment environment) {
+        this.environment = environment;
+    }
+
+    /**
+     * Registers a new {@link AnalysisProvider} to create
+     * {@link CharFilterFactory} instance per node as well as per index.
+     */
+    public void addCharFilter(String name, AnalysisProvider<CharFilterFactory> charFilter) {
```

Comment:
I think we've have mixed use of "add" and "register". Should we stick with "register" as I think that is what we have used most recently?